java/core/src/test/java/org/bondlib/StructGenerator.java

```diff
@@ -0,0 +1,408 @@
+            returnValue = setValue;
+        } else if (bondTypeClass == MapBondType.class) {
+            // map<,>
+            MapBondType mapBondType = (MapBondType) bondType;
+            BondType mapKeyBondType = mapBondType.getKeyType();
+            BondType mapValueBondType = mapBondType.getValueType();
+            Map mapValue = mapBondType.newDefaultValue();
+            this.generateMapElements(mapValue, mapKeyBondType, mapValueBondType, context);
+            returnValue = mapValue;
+        } else if (bondTypeClass == BondedBondType.class) {
+            // bonded<>
+            BondedBondType bondedBondType = (BondedBondType) bondType;
+            StructBondType declaredStructBondType = bondedBondType.getValueType();
+            ArrayList<StructBondType> allowedPolymorphicStructBondTypes =
+                    this.getPolymorphicStructBondTypesFor(declaredStructBondType);
+            int index = this.rand.nextInt(allowedPolymorphicStructBondTypes.size());
+            StructBondType actualStructBondType = allowedPolymorphicStructBondTypes.get(index);
+            BondSerializable structValue = this.generateStruct(actualStructBondType, context);
+            returnValue = Bonded.fromObject(structValue);
+        } else if (EnumBondType.class.isAssignableFrom(bondTypeClass)) {
+            // generated enum
+            EnumBondType enumBondType = (EnumBondType) bondType;
+            returnValue = this.generateEnumValue(enumBondType);
+        } else if (StructBondType.class.isAssignableFrom(bondTypeClass)) {
+            StructBondType structBondType = (StructBondType) bondType;
+            returnValue = this.generateStruct(structBondType, context);
+        } else {
+            // unknown
+            throw new IllegalArgumentException("Unknown BondType implementation: " + bondTypeClass);
+        }
+        return returnValue;
+    }
+
+    // returns true if a pseudo-random number distributed uniformly between 0 and 1 is less than the argument
+    private boolean testProbability(double probability) {
+        return this.rand.nextDouble() < probability;
+    }
+
+    // helper to generate a pseudo-random integer with the given number of bits (optionally signed)
+    private long generateIntValue(int bitCount, boolean isSigned) {
+        // Given number N randomly chosen between 1 and bitCount with uniform probability, determine the
+        // magnitude of the result as an unsigned value of N bits where bits 0 to N-1 are randomly chosen
+        // individually (with uniform probability), and all the other bits are set to 0. Then optionally
+        // negate the result with 50% probability.
+        // This algorithm improves coverage of smaller integers for variable-length encoding.
+        int n = this.rand.nextInt(bitCount) + 1;
+        long result = 0;
+        for (int i = 0; i < n; ++i) {
+            if (this.rand.nextBoolean()) {
+                result |= (1 << i);
+            }
+        }
```

Comment:
Any reason this can't be
```
int resultBits = this.rand.nextInt(bitCount) + 1;
long result = this.rand.nextInt(1 << resultBits);
```
?

Reply:
Won't work cleanly for unsigned. The current code relies on bit manipulations rather than signed arithmetic, which does the trick.

---
In reply to: [138759100](https://github.com/Microsoft/bond/pull/607#discussion_r138759100) [](ancestors = 138759100)

Reply:
Not sure I understand, but it's fine as is.